File: lib/fftfreq.m
function f =  fftfreq(N,ts)
%FFTFREQ computes the positive and negative frequencies sampled in the FFT
%
% Syntax:   f = fftfreq(N,ts)
%
% Description:
%
% FFTFREQ computes a column vector of N frequencies with sampling frequency
% 1/ts, with circular frequencies mapped to positive and negative
% frequencies that are centered at zero.
%
% Inputs:
%   N   Number of time samples
%   ts  Sampling time
%
% Outputs:
%   f   Frequency vector (1/ts)
%

validateattributes(N,{'numeric'},{'scalar','positive'})
validateattributes(ts,{'numeric'},{'scalar','positive'})

fs = 1/(ts*N);

% Use IFFTSHIFT to shift zero frequency from center to first position.
kcyclic = ifftshift((0:N-1)-floor(N/2));

f = kcyclic*fs;
f = f(:);
    
end

File: thztoolsPY/fftfreq.py
import numpy as np


def fftfreq(n, t):
    """
    Computes the positive and negative frequencies sampled in the Fast Fourier Transform.

    Parameters
    ----------
    n : int
        Number of time samples
    t: float
        Sampling time

    Returns
    -------
    f : ndarray
        Frequency vector (1/ts) of length n containing the sample frequencies.

    """

    if n % 2 == 1:
        f = np.fft.fftfreq(n, t)
    else:
        f = np.fft.fftfreq(n, t)
        f[int(n/2)] = -f[int(n/2)]

    return f
